### DISCUSSION TITLE
Unused variables warning in .env files

### DISCUSSION THREAD

**Discussion by RHOOPH** [LazyVim/LazyVim#4027] [Q&A]
At first, only `.env` files were syntax highlighted, and not `.env.production` or `.env.local`, So i enabled  Dot Files extras. 
Now syntax is properly highlighted in all files but I get `<variable-name> appears unused. Verify use (or export if used externally).` as warning. (Edit: These warnings were there even before dot file extras are enabled)

Lsp info
```
 Detected filetype:   sh
 
 1 client(s) attached to this buffer: 
 
 Client: bashls (id: 1, bufnr: [5])
 	filetypes:       sh
 	autostart:       true

```
sample file :arrow_down: 
```
GOOGLE_CLIENT_ID=hidden
GOOGLE_CLIENT_SECRET=hidden
NEXTAUTH_SECRET=hidden
NEXTAUTH_URL=hidden
```
how to highlight files without getting these warnings ?

**Comment by apaydev** [comment]
I don't know if it's a good practice, but I just disable the warnings for that specific exception with the shellcheck comment. https://www.shellcheck.net/wiki/SC2034
```
# shellcheck disable=SC2034
GOOGLE_CLIENT_ID=hidden
GOOGLE_CLIENT_SECRET=hidden
NEXTAUTH_SECRET=hidden
NEXTAUTH_URL=hidden
```

**Comment by cbaconnier** [comment]
This gave me crazy and I fixed this this way:

![image](https://github.com/user-attachments/assets/84b14379-5938-4f7c-bd30-04ab8c615050)


I forced the file type to `env`instead of `sh`
**nvim/plugins/filetype.lua**
```lua
return {
  {
    "nathom/filetype.nvim",
    lazy = false,
    opts = {
      overrides = {
        extensions = {
          env = "env",
        },
        complex = {
          [".env.*"] = "env",
        },
      },
    },
  },
}
```
Unfortunately, the syntax highlighting won't work anymore.
So I have added this syntax highlighting file (Thanks to https://github.com/overleaf/vim-env-syntax):  
**nvim/after/syntax/env.vim**
```lua
" https://github.com/overleaf/vim-env-syntax/blob/main/syntax/env.vim
" Vim syntax file
" Language:		env files (NOT shell code)
" Maintainer:		Gernot Schulz <gernot.schulz@overleaf.com>

" quit when a syntax file was already loaded
if exists("b:current_syntax")
  finish
endif

syn match  envComment				'^#.*'

syn match  envVariableUnassigned		"^\<\h\w*$"

syn match  envVariable				"^\<\h\w*\ze="	nextgroup=envVarAssign
syn match  envVarAssign		contained	"="		nextgroup=envVar,envQuotedVarOpen,envSpace
syn match  envVar		contained	"\h\w*"
syn match  envSpace		contained	"\s\+"		nextgroup=envQuotedVarOpen
syn match  envQuotedVarOpen	contained	"[\"']"

syn match  envSpace				"\s$"
syn match  envQuotedVarClose			"[\"']$"


let b:current_syntax = "env"

hi def link envComment			Comment

hi def link envVariableUnassigned	Error

hi def link envVariable			Identifier
hi def link envVarAssign		Operator
hi def link envSpace			Error
hi def link envQuotedVarOpen		Error
hi def link envQuotedVarClose		Error

```

Finally we apply with:
**nvim/lua/config/autocmds.lua**
```
vim.api.nvim_create_autocmd({ "BufNewFile", "BufRead" }, {
  pattern = ".env*",
  callback = function()
    vim.bo.filetype = "env"
  end,
})
```

_Note: I'm pretty sure the filetype plugin is not necessary, but I couldn't make it work without autocmd nor autocmd without the filetype plugin._

**Comment by loqusion** [comment]
Changing the `'filetype'` of `.env*` files is unnecessary. You can disable diagnostics for just `.env*` files and still enjoy the benefits of syntax highlighting, completion, formatting, etc. by dropping this into your config:

```lua
  {
    "neovim/nvim-lspconfig",
    opts = function(_, opts)
      local on_publish_diagnostics = vim.lsp.diagnostic.on_publish_diagnostics
      opts.servers.bashls = vim.tbl_deep_extend("force", opts.servers.bashls or {}, {
        handlers = {
          ["textDocument/publishDiagnostics"] = function(err, res, ...)
            local file_name = vim.fn.fnamemodify(vim.uri_to_fname(res.uri), ":t")
            if string.match(file_name, "^%.env") == nil then
              return on_publish_diagnostics(err, res, ...)
            end
          end,
        },
      })
    end,
  },
```

**Reply by cbaconnier** [reply]
Thanks ! I first try something similar, but I couldn't get it work. Your solution is working like a charm.

Note : for my usage, I also had to force the filetype to `sh` for the files like `.env.example` or `.env.testing`, otherwise it doesn't recognize them.

```lua
vim.api.nvim_create_autocmd({ "BufNewFile", "BufRead" }, {
  pattern = ".env*",
  callback = function()
    vim.bo.filetype = "sh"
  end,
})

```

**Reply by jleweli** [reply]
@loqusion Your solution worked for me, but somehow stopped working a few days ago. I think there might has been an update in the meantime which prevents the solution from working.

I solved the problem for me by removing the lazy extra [util.dot](https://www.lazyvim.org/extras/util/dot) as well as the lua file with your code snippet inside. Then, I removed the bash-language-server (bashls) via Mason.

Now, I have no more warnings in my .env files. Leaving this here for other users who are new to LazyVim and have the same problem.

**Comment by tom-gora** [comment]
I tweaked this into even a simpler approach. In my setup I call config function for `lspconfig` plugin. In there I have a standard loop calling setup() on all declared lsps, but the twist is I inject conditional doing custom tweaks to any server that might need it (else just simple call to setup)
```lua
	config = function()
	-- .... various declarations, bring in server list from mason
        local lsps = mason_conf.mason_lspconfig.ensure_installed
	
	for _, lsp in ipairs(lsps) do
		if lsp == "bashls" then
			-- disable annoying diagnostic noise in .env files. those are not bash scripts
			local file_name = vim.fs.basename(vim.api.nvim_buf_get_name(0))
                        -- brainded implementation if filename matches the string just call  func to disable diagnostic. no digging into plugin api or anything like that
			if string.match(file_name, "^%.env") == nil then
				vim.diagnostic.enable(false)
			end
			-- make bashls attach to additional filetypes
			lspconfig.bashls.setup({
				capabilities = capabilities,
				filetypes = { "sh", "zsh" },
			})
		elseif lsp == "emmet_language_server" then
		--- ....
                --- other custom tweaks I might do on per-lsp basis
          
                --- else just call setup and call it a night. there. done and it works
		else
			lspconfig[lsp].setup({ capabilities = capabilities })
		end
	end

```
EDIT: Kudos to @loqusion for putting me on the right track with suggesting filename patternmatch - simple solutions always the best.

**Reply by loqusion** [reply]
Unless I'm mistaken, this code disables diagnostics globally if the filename of the "current buffer" (buffer 0) when the `config` function is run (i.e. when `lspconfig.nvim` is loaded) matches `^%.env`, right? I assume that's not what you want.

**Reply by tom-gora** [reply]
@loqusion You're right. Missed that when swapping from deprecated disable() to enable(false). The old way took 0 to apply only to the current buffer. I'm sure the changed API has a way to scope as well. Will look into it tomorrow when back at the keyboard. Thanks for pointing that out!

**Reply by tom-gora** [reply]
After some tinkering I confirmed that trying to disable this way inside of lspconfig fails with filter `{ bufnr = 0}` added. However I still preferred using built in diag functions rather than hooking into plugin opts with custom handlers. COmbining thread suggestions I just used an autocmd grabbing precise bufnr from event and it works now ok:
```lua
autocmd({ "BufNewFile", "BufReadPost" }, {
	group = lsp_hacks,
	pattern = ".env*",
	callback = function(e)
		vim.diagnostic.enable(false, { bufnr = e.buf })
	end,
})
```
Thanks for great feedback and suggestions :)

**Reply by valerykh** [reply]
> After some tinkering I confirmed that trying to disable this way inside of lspconfig fails with filter `{ bufnr = 0}` added. However I still preferred using built in diag functions rather than hooking into plugin opts with custom handlers. COmbining thread suggestions I just used an autocmd grabbing precise bufnr from event and it works now ok:
> 
> ```lua
> autocmd({ "BufNewFile", "BufReadPost" }, {
> 	group = lsp_hacks,
> 	pattern = ".env*",
> 	callback = function(e)
> 		vim.diagnostic.enable(false, { bufnr = e.buf })
> 	end,
> })
> ```
> 
> Thanks for great feedback and suggestions :)

Thanks. I put this in `autocmds.lua` as workarond for now and it works fine
```lua
local lsp_hacks = vim.api.nvim_create_augroup("LspHacks", { clear = true })

vim.api.nvim_create_autocmd({ "BufNewFile", "BufReadPost" }, {
  group = lsp_hacks,
  pattern = ".env*",
  callback = function(e)
    vim.diagnostic.enable(false, { bufnr = e.buf })
  end,
})
```

**Reply by barrettruth** [reply]
This will not work when opening neovim with `.env`. the following did for me:

```lua
    on_attach = function(_, bufnr)
        local bufname = vim.api.nvim_buf_get_name(bufnr)
        if string.match(bufname, '%.env') then
            vim.diagnostic.enable(false, { bufnr = bufnr })
        end
    end,
```
